File: backends/dotnet/shared/Implementations/Middleware/Idempotency.Default/IdempotencyOptions.cs
// -----------------------------------------------------------------------
// <copyright file="IdempotencyOptions.cs" company="DCSV">
// Copyright (c) DCSV. All rights reserved.
// </copyright>
// -----------------------------------------------------------------------

namespace D2.Shared.Idempotency.Default;

/// <summary>
/// Configuration options for idempotency middleware.
/// </summary>
public class IdempotencyOptions
{
    /// <summary>
    /// Gets or sets the TTL for cached responses.
    /// </summary>
    /// <remarks>
    /// Default is 24 hours. After this period, the same idempotency key can produce a new response.
    /// </remarks>
    public TimeSpan CacheTtl { get; set; } = TimeSpan.FromHours(24);

    /// <summary>
    /// Gets or sets the TTL for the in-flight sentinel lock.
    /// </summary>
    /// <remarks>
    /// Default is 30 seconds. Acts as a safety valve if the server crashes mid-request.
    /// </remarks>
    public TimeSpan InFlightTtl { get; set; } = TimeSpan.FromSeconds(30);

    /// <summary>
    /// Gets or sets the maximum response body size (in bytes) that will be cached.
    /// </summary>
    /// <remarks>
    /// Default is 1 MB. Responses exceeding this size will not be cached.
    /// </remarks>
    public int MaxBodySizeBytes { get; set; } = 1_048_576;

    /// <summary>
    /// Gets or sets the HTTP methods that the idempotency middleware applies to.
    /// </summary>
    /// <remarks>
    /// Default includes POST, PUT, PATCH, DELETE. GET, HEAD, and OPTIONS are always skipped.
    /// </remarks>
    public HashSet<string> ApplicableMethods { get; set; } =
        [HttpMethod.Post.Method, HttpMethod.Put.Method, HttpMethod.Patch.Method, HttpMethod.Delete.Method];

    /// <summary>
    /// Gets or sets a value indicating whether error responses (4xx/5xx) should be cached.
    /// </summary>
    /// <remarks>
    /// Default is false. When false, non-2xx responses remove the sentinel so clients can retry.
    /// </remarks>
    public bool CacheErrorResponses { get; set; }
}

File: backends/node/shared/implementations/middleware/idempotency/default/src/idempotency-options.ts
/** Configuration options for idempotency middleware. */
export interface IdempotencyOptions {
  /** TTL for cached responses in milliseconds. Default: 24 hours. */
  cacheTtlMs: number;
  /** TTL for the in-flight sentinel lock in milliseconds. Default: 30 seconds. */
  inFlightTtlMs: number;
  /** Maximum response body size in bytes that will be cached. Default: 1 MB. */
  maxBodySizeBytes: number;
  /**
   * HTTP methods that idempotency applies to.
   * Requests with other methods skip the idempotency check entirely.
   * Default: POST, PUT, PATCH, DELETE.
   */
  applicableMethods: ReadonlySet<string>;
  /**
   * Whether error responses (4xx/5xx) should be cached.
   * When false, non-2xx responses remove the sentinel so clients can retry.
   * Default: false.
   */
  cacheErrorResponses: boolean;
}

export const DEFAULT_IDEMPOTENCY_OPTIONS: IdempotencyOptions = {
  cacheTtlMs: 86_400_000, // 24 hours
  inFlightTtlMs: 30_000, // 30 seconds
  maxBodySizeBytes: 1_048_576, // 1 MB
  applicableMethods: new Set(["POST", "PUT", "PATCH", "DELETE"]),
  cacheErrorResponses: false,
};
